doc/compiler/optimizer/OptimizerDesignFeatures.md

```diff
@@ -0,0 +1,303 @@
+extended basic block is represented by a single node regardless of how many consumers (parent nodes in the
+OpenJ9 JIT IL representation) of the value there are in that extended basic block.
+
+It also makes the concept of "value equality" trivial for most local analyses, since this basically just
+boils down to "node equality" making the code much cleaner and obviates the need for a separate analysis
+step that tracks "value equality".
+
+Another aspect of the OpenJ9 JIT IL representation that is novel is the use of exception check opcodes
+and exception edges in the Control Flow Graph (CFG).
+
+Together these two notions avoid chopping up the compiled method into many small basic blocks ending every
+time an exception check must be done.
+
+Essentially exception check opcodes represent control flow exiting the extended basic block going to the
+exception catch block (if there exists one in the method) and exception edges represent this flow in the CFG.
+
+Crucially this is done without ending the basic block every time an exception check is encountered, and the
+"special" nature of exception checks and edges is known to the compiler's generic
+[dataflow analysis](https://youtu.be/YCCdJ1Qphao) framework, so that it does the right operations every time
+those special idioms are encountered.
+
+This decreases the complexity of the CFG in general and makes the control flow look largely like it would have
+done for a language without exception checking (e.g., C) and restricts the area of complexity in the compiler
+to the dataflow analysis framework that is only changed very rarely in practice.
+
+## Cooperative Threading Model and Compiled Method Meta data
+
+Dynamic languages that use JIT compilers often have other requirements such as support for garbage collection (GC),
+exception stack trace, debug code and hot code replacement.
+
+In most of these cases, there is a need to update some "state" as execution proceeds to support the desired
+functionality, while at the same time, the desired functionality is only needed in rare instances (e.g., a user
+wanting to do a debug session is rare as is the need for an exception stack trace usually).
+
+A naive solution might be to update the "state" as execution proceeds on the main line path, but this has the downside
+that there is a lot of work done on the main line path that may never (or rarely) in fact be used. Most of these
+requirements really hinge around locating the "state" needed to support the functionality at a given program point.
+
+The key insight here is that this problem can be solved more efficiently by a) restricting the program points where
+such "state" could be needed, b) forcing threads to only yield at these restricted program points and employing a
+stack walker to figure out where each thread is, and c) locate the meta data (capturing the "state") laid down by
+the JIT at compile time for each of these restricted program points.
```

Comment:
Maybe this would be easier to read as a bulleted list rather than using "a), b), c)"